Add new api endpoint for AIP Migration

jhsimpson opened the pull request:
This PR is meant to replace https://github.com/artefactual/archivematica-storage-service/pull/130 which is based on 0.7.x.

This PR brings in a new feature, allowing AIPs to be moved from one Location to another.  This is done though a new API endpoint.

This also brings in celery and redis as new dependencies.

Hwesta committed:
HACK remove this - disable SSL checking

mcantelon committed:
Added package move logic and endpoint. refs #9909

Add method to package to change the package's storage location. Add REST
endpoint for triggering move to execute as a background process.

mcantelon committed:
Startup fix to package move statuses. refs #9909

When storage service starts, look for packages with status "MOVING" and
change the status to "MOVE_FAILED". This is done to allow for the possibilty
of moves being interrupted by power failures, etc.

mcantelon committed:
Update pointer file after AIP move. refs #9909

After a package's storage location is changed, update the package's pointer
file if applicable.

mcantelon committed:
Fix package filter issue, add sorting. refs #9909

Add ability to filter by current location. Add ability to order by UUID.

mcantelon committed:
Fixed/simplified exception call. refs #9909

mcantelon committed:
Add Celery for async move.

scollazo committed:
Update version to 0.9.1

Hwesta commented on storage_service/locations/models/package.py:
```diff
@@ -317,6 +319,67 @@ def _update_quotas(self, space, location):
+            origin_space = self.current_location.space
+            origin_space.move_to_storage_service(
+                source_path=source_path,
+                destination_path='move',
```
This should probably be a temp directory, so multiple AIPs being moved don't interfere with each other.

hakamine replied:
also the destination path 'move' seems not being created elsewhere, so the rsync triggered `move_to_location()` produces an error if the directory `/var/archivematica/storage_service/move` was not manually created beforehand.

Hwesta commented on storage_service/locations/models/package.py:
```diff
@@ -317,6 +319,67 @@ def _update_quotas(self, space, location):
+
+            # Update location
+            self.current_location = destination_location
+            self.status = Package.UPLOADED
```
Do we know this is the correct status for all Spaces?

Hwesta commented on storage_service/locations/models/package.py:
```diff
@@ -317,6 +319,83 @@ def _update_quotas(self, space, location):
+
+        except Exception:
+            LOGGER.info('Attempt to move package %s to location %s failed', self.uuid, destination_location_uuid, exc_info=True)
+            self.status = Package.MOVE_FAILED
```
Do we need a `MOVE_FAILED` status or can we just display an error message in the GUI? Should it be set back to the previous state? Do we do different behaviour if the package failed move?

Hwesta commented on storage_service/locations/api/resources.py:
```diff
@@ -592,6 +595,30 @@ def recover_aip_request(self, request, bundle, **kwargs):
+        if package.status == Package.MOVING:
+            message = "Move already in progress."
+            success = False
+        else:
```
Do we want to move on every package status?  Probably not for Failed, Deleted, Pending (not yet on SS), maybe not for Delete requested or recovery request.

hakamine replied:
@Hwesta @jhsimpson would it be too restrictive to allow only moves of packages in the UPLOADED state? Asking because the code seems to be setting the state to UPLOADED if the move was successful, so probably this state was assumed? ( https://github.com/artefactual/archivematica-storage-service/pull/143/files#diff-5f2ead45b8622f27e9b0bf6333334fffR380 )

Hwesta commented on storage_service/locations/models/package.py:
```diff
@@ -317,6 +319,83 @@ def _update_quotas(self, space, location):
+        """
+        Move a package to another location (that has the same purpose as package's current location).
+        """
+        destination_location = Location.active.get(uuid=destination_location_uuid)
```
Should this be checked in `resources.move_request`? It's user-submitted data, we should verify the location exists before trying to start a move and immediately erroring.

Hwesta commented on storage_service/locations/api/resources.py:
```diff
@@ -592,6 +595,30 @@ def recover_aip_request(self, request, bundle, **kwargs):
+
+        response_data = {'success': success, 'message': message}
+
+        return http.HttpResponse(
```
Since this starts an async process, should it return 201 now?

Hwesta commented on README.rst:
```diff
@@ -1 +1,5 @@
+
+Start worker:
+
+celery -A storage_service worker -l info
```
Has the ansible role been updated to start celery too?

jhsimpson replied:
Not yet.  Updating the archivematica ansible role is going to be addressed in this ticket:
https://projects.artefactual.com/issues/10285, which is assigned to hakamine.  

That ticket will blocks merging this PR, I think.

hakamine replied:
This is being done in this branch of the ansible role (not merged to master yet) https://github.com/artefactual-labs/ansible-archivematica-src/tree/dev/ss-celery-worker

hakamine replied:
ansible role updated, merged to master branch  https://github.com/artefactual-labs/ansible-archivematica-src/pull/92

hakamine committed:
Move packages: fixes (SSPR #143)

 - Create a temporary directory with random naming for package moving,
   instead of using a fixed-name one, to avoid concurrency conflicts

 - Use content_type in http response instead of mimetype in
   move_request() HttpResponse

 - If move_to_location() fails, revert to original package status
   instead of setting it to MOVE_FAILED

 - Allow to move packages only if in UPLOADED state

hakamine committed:
HACK: do not count transfer source objects

Do not count transfer source objects when browsing to improve UI
responsiveness.

sevein commented:
I'm trying to think on not so ideal scenarios where web or task workers hang forever or exit unexpectedly, or race conditions in general in the context of this feature. Is the code safe enough? There are things like [database transactions](http://celery.readthedocs.io/en/latest/userguide/tasks.html#database-transactions), Celery's task [result/state backend](http://celery.readthedocs.io/en/latest/userguide/tasks.html#result-backends), custom task identification (e.g. `task_id` named after package UUID for easy discovery), use of `CELERY_TRACK_STARTED`, handlers ([on_success](http://celery.readthedocs.io/en/latest/userguide/tasks.html#on_success), [on_failure](http://celery.readthedocs.io/en/latest/userguide/tasks.html#on_failure)), timeouts, retries, [global locking](http://loose-bits.com/2010/10/distributed-task-locking-in-celery.html) and a few of other things that we should probably consider before we move ahead to guarantee data integrity. I find it can be overwhelming but should we at least try to cover the basics?

There is a list with links to projects using Celery at their wiki: https://github.com/celery/celery/wiki#projects--applications-usingfor-celery. It's hard to find examples that are not too basic, on the other hand the [pulp](https://github.com/pulp/pulp) project looks very sophisticated.

mcantelon committed:
Added debug logging of source/dest for os.rename instance in space model.

hakamine committed:
Fix logger in space model

mcantelon committed:
Fix issue moving between spaces, refs: #9909

Because the destination space wasn't being specified, the move to processing
space would put the package in the wrong processing space and the move from
wouldn't be able to find the file(s).
